### PR TITLE
Remove DynamicData name mapping APIs

### PR DESCRIPTION
**Pull request by annelo-msft**
# Description
This PR removes name mapping APIs, per discussion below.

Addresses https://github.com/Azure/azure-sdk-for-net/issues/35944

- [ ] Rework DynamicDataOptions.Default and DynamicDataNameMapping enum such that
  - [ ] anything called Default has the same behavior as not passing any parameteters
  - [ ] a good default experience works for both read/write of the JSON
  - [ ] enum names make it clear what the resulting behavior of the type will be
- [ ] Remove one of the ToDynamicFromJson() overloads, contingent on outcome of DynamicDataNameMapping discussion
- [ ] Keep DynamicDataOptions in the root to enable passing them frequently?

# Prior Discussion

## Problem

Per [comments on the APIView](https://apiview.dev/Assemblies/Review/77bdbcef49a3465fa96fac3fb745bf57/759e2571c2fc408d89784be8152f3169?diffRevisionId=047e602a417146e8a83b835306a2fc42&doc=False&diffOnly=True#Azure.DynamicDataNameMapping) around `DynamicDataNameMapping`, and open questions around how we set defaults, here is one proposal for updates to the API.

### Read/Write == Get/Set == Deserialize/Serialize
image

### Mapping is from outer to inner not to getters or setters
image

### CODE REVIEW

File: sdk/core/Azure.Core/api/Azure.Core.net461.cs [hunk 1]
```diff
@@ -811,11 +811,16 @@ public sealed partial class DynamicData : System.Dynamic.IDynamicMetaObjectProvi
         System.Dynamic.DynamicMetaObject System.Dynamic.IDynamicMetaObjectProvider.GetMetaObject(System.Linq.Expressions.Expression parameter) { throw null; }
         public override string ToString() { throw null; }
     }
+    public enum DynamicDataDefaults
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/api/Azure.Core.net461.cs:814]: This is an approach used by JsonSerializer, e.g. [JsonSerializerDefaults](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializerdefaults?view=net-8.0), and by [ASP.NET Core for web apps](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/configure-options?pivots=dotnet-7-0#web-defaults-for-jsonserializeroptions)

**Comment by annelo-msft** [comment]
Per discussion in the .NET Azure.Core stand-up, we will do auto-mappings according to the behavior described by `PascalCaseDynamicCamelCaseData`.  This should work for the vast majority of scenarios, since camelCase is stylistically correct for JSON and PascalCase is stylistically correct for CSharp.

We rationalize the difference in this behavior from types like JsonDocument or serialized anonymous types, because those are types that work on string literals.  The design goal of the `DynamicData` type is to serve as a bridge between CSharp and JSON, and operate in a manner similar to Azure SDK model types in the majority of cases.

The next iteration of this PR will remove the options APIs entirely.

File: sdk/core/Azure.Core/samples/DynamicContent.md [hunk 2]
```diff
@@ -96,7 +96,7 @@ When working with JSON from Azure services, you can learn what properties are av
 
 ```C# Snippet:AzureCoreUseDynamicDataDefaults
 Response response = client.GetWidget();
-dynamic widget = response.Content.ToDynamicFromJson(DynamicDataOptions.Default);
+dynamic widget = response.Content.ToDynamicFromJson(new DynamicDataOptions(DynamicDataDefaults.Azure));
```
[the remark this answers is not in the document]
Comment (annelo-msft) [reply sdk/core/Azure.Core/samples/DynamicContent.md]: Addressed per discussion.

File: sdk/core/Azure.Core/api/Azure.Core.net461.cs (hunk 1 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/api/Azure.Core.net461.cs]: Removed per discussion.

File: sdk/core/Azure.Core/samples/DynamicContent.md [hunk 3]
```diff
@@ -90,27 +92,13 @@ dynamic widget = response.Content.ToDynamicFromJson();
...
-
-```C# Snippet:AzureCoreUseDynamicDataDefaults
-Response response = client.GetWidget();
-dynamic widget = response.Content.ToDynamicFromJson(DynamicDataOptions.Default);
-string id = widget.Id;
-widget.Name = "New Name";
-client.SetWidget(RequestContent.Create(widget));
-```
-
-If no parameter is passed to `ToDynamicFromJson()`, properties names must exactly match the member names in the JSON content.  Passing `DynamicDataOptions.Default` will enable properties to be accessed using PascalCase property names, and will write any added properties with camelCase names.
```
Comment (KrzysztofCwalina) [inline sdk/core/Azure.Core/samples/DynamicContent.md:105]: Should we have a note at the end of this document explaining how the default mapping works?

**Review by heaths — COMMENTED** [review]
@KrzysztofCwalina I still have concerns about the impact of this chnage. It means that models will have different casing going in (anonymous types) vs. out and that seems really confusing. Why use different case for different scenarios? If we're going to expose JSON bindings, why not just leave the case unchanged like any other languages are doing? Since a customer has to use (normally) camelCase for input, why would they not expect to access the result as camelCase instead of PascalCase?

In reverse, if they've initially worked only with response models and now have to write an input model, why would they not think to use PascalCase and get some cryptic service error about a missing member they "knew" they added (just the wrong case)? I don't think this is a friendly UX.

In the case of CLU here, the only reason I used PascalCase is because I did the conversion from having models to dynamic, so I was able to leave the case as-is and they were only for tests, not samples.

**Comment by annelo-msft** [comment]
@heaths, would you be able to provide some of the specific use cases you're concerned about?

**Comment by heaths** [comment]
We've definitely seen some, and maybe @tg-msft remembers (I'm looking), but even as @KrzysztofCwalina said in your sync-up meeting recording, "99% should be camel case". That may well be, but in that 1% (or even less) we're responsible for improper casing and potential data loss.

And what about special cases, like if a customer sets `model.PIICategories`? How do you know to change that to `piiCategories` that Cognitive requires? What about `IPAddress` vs. `ipAddress`? Now you're also having to implement all these "rules" about how to map things, which also may not be consistent.

The different with the customer making a typo in any property vs. us casing it incorrectly is that we did the latter through no fault of the customer.

At least if they have a way to control the case themselves and we're clear about it - like does https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/DynamicContent.md#get-a-property-with-invalid-c-characters-in-the-name give them a way - they can override our case changing. Does it? Maybe we should explicitly call out that case so it's searchable.

File: sdk/core/Azure.Core/samples/DynamicContent.md (hunk 3 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/samples/DynamicContent.md:108]: Added.

**Comment by johanste** [comment]
For inspiration, this is what we do in the Python code generator:

- We return what is (if I squint a bit) the dynamic object directly from the client operation. There is no separate GetDynamic call needed.
- When accessing properties of the returned value by key name (e.g. `result["someProperty"]`, we use the casing in the raw data. We also return the raw data (e.g. no type conversations from string -> date-time for date time typed properties. We never do case or snake conversion when accessing by key name.
- We only provide property accessors as part of having a model. We actually codegen properties in this case rather than using dynamic runtime hooks in order to:
   1. Hang documentation off something
   2. Get IntelliSense and static type checking in IDEs that support it.
- When accessing values through property accessors, we also do type conversion.

```python
# Assuming we have generated "models" (property names) for the library/do_something method:
result = client.do_something() # returns { "someProperty": "hello", "someDateTime": "2023-01-01T01:02:03Z" }

# A string is a string and can be accessed either way.
assert result['someProperty'] == result.some_property
assert result['someDateTime'] == "2023-01-01T01:02:03Z"
assert isinstance(result.some_date_time, datetime.datetime)

try:
    result['some_property']
    assert False # there is no such key...
except KeyError:
    # happy

try:
    result.someProperty
    assert False # there is no such property...
except AttributeError:
    # happy
```

**Review by christothes — APPROVED** [review]

**Review by jsquire — APPROVED** [review]

**Pull request merged**